numpy/_core/src/umath/matmul.c.src

```diff
@@ -515,3 +515,165 @@ NPY_NO_EXPORT void
+CLONGDOUBLE_dotc(char *ip1, npy_intp is1, char *ip2, npy_intp is2,
+                 char *op, npy_intp n, void *NPY_UNUSED(ignore))
+{
+    npy_longdouble tmpr = 0.0L;
+    npy_longdouble tmpi = 0.0L;
+    npy_intp i;
+
+    for (i = 0; i < n; i++, ip1 += is1, ip2 += is2) {
+        const npy_longdouble ip1r = ((npy_longdouble *)ip1)[0];
+        const npy_longdouble ip1i = ((npy_longdouble *)ip1)[1];
+        const npy_longdouble ip2r = ((npy_longdouble *)ip2)[0];
+        const npy_longdouble ip2i = ((npy_longdouble *)ip2)[1];
+        /* ip1.conj() * ip2 */
+        tmpr += ip1r * ip2r + ip1i * ip2i;
+        tmpi += ip1r * ip2i - ip1i * ip2r;
+    }
+    ((npy_longdouble *)op)[0] = tmpr;
+    ((npy_longdouble *)op)[1] = tmpi;
+}
+
+static void
+OBJECT_dotc(char *ip1, npy_intp is1, char *ip2, npy_intp is2, char *op, npy_intp n,
+            void *NPY_UNUSED(ignore))
+{
+    npy_intp i;
+    PyObject *tmp1, *tmp2, *tmp = NULL;
+    PyObject **tmp3;
+    for (i = 0; i < n; i++, ip1 += is1, ip2 += is2) {
+        if ((*((PyObject **)ip1) == NULL) || (*((PyObject **)ip2) == NULL)) {
+            tmp1 = Py_False;
+            Py_INCREF(Py_False);
+        }
+        else {
+            PyObject *obj1conj = PyObject_CallMethod(*((PyObject **)ip1),
+                                                     "conjugate", NULL);
+            if (!obj1conj) {
+                Py_XDECREF(tmp);
+                return;
+            }
+            tmp1 = PyNumber_Multiply(obj1conj, *((PyObject **)ip2));
+            Py_DECREF(obj1conj);
+            if (!tmp1) {
+                Py_XDECREF(tmp);
+                return;
+            }
+        }
+        if (i == 0) {
+            tmp = tmp1;
+        }
+        else {
+            tmp2 = PyNumber_Add(tmp, tmp1);
+            Py_XDECREF(tmp);
+            Py_XDECREF(tmp1);
+            if (!tmp2) {
+                return;
+            }
+            tmp = tmp2;
+        }
+    }
+    tmp3 = (PyObject**) op;
+    tmp2 = *tmp3;
+    *((PyObject **)op) = tmp;
+    Py_XDECREF(tmp2);
+}
+
+/**begin repeat
+ *  #TYPE = FLOAT, DOUBLE, LONGDOUBLE, HALF,
+ *          UBYTE, USHORT, UINT, ULONG, ULONGLONG,
+ *          BYTE, SHORT, INT, LONG, LONGLONG, BOOL,
+ *          CFLOAT, CDOUBLE, CLONGDOUBLE, OBJECT#
+ *  #DOT = dot*15, dotc*4#
+ */
+
+NPY_NO_EXPORT void
+@TYPE@_vecdot(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
+{
+    npy_intp dOuter = *dimensions++;
+    npy_intp iOuter;
+    npy_intp s0 = *steps++;
+    npy_intp s1 = *steps++;
+    npy_intp s2 = *steps++;
+    npy_intp dn = dimensions[0];
+    npy_intp is1_n=steps[0], is2_n=steps[1];
+
+    for (iOuter = 0; iOuter < dOuter; iOuter++,
+                         args[0] += s0, args[1] += s1, args[2] += s2) {
+        void *ip1=args[0], *ip2=args[1], *op=args[2];
+        /*
+         * TODO: refactor this out to an inner_loop_selector?
+         */
+        @TYPE@_@DOT@(ip1, is1_n, ip2, is2_n, op, dn, NULL);
```

Comment:
Ah, missed one thing...

This is annoying, because we really should have an error return, for object, the error paths are not particularly well defined (might do more operations).

It needs moving to the new API to fix, so maybe you can add a `PyErr_Occured()` check here for object (as much as I dislike it)?

Reply:
OK, done (and test added)